### PR TITLE
Begin work in making sure `request.state` is preserved between middlewares.

### PR DESCRIPTION
**Pull request by didip**
One test still fails, and `request.state` is still not preserved but I'd like to share to you the code. I am not clear why the `scope` always empty on every new `Request`.

See issue #551 

cc @tomchristie

### CODE REVIEW

File: starlette/requests.py [hunk 1]
```diff
@@ -32,6 +33,9 @@ def __init__(self, scope: Scope, receive: Receive = None) -> None:
         assert scope["type"] in ("http", "websocket")
         self._scope = scope
 
+        # Ensure 'state' has an empty dict if it's not already populated.
+        self._scope.setdefault('state', {})
```
Comment (lovelydinosaur) [inline starlette/requests.py]: Let's *only* do this if/when the `.state` property is first accessed.

File: starlette/requests.py [hunk 2]
```diff
@@ -19,7 +19,8 @@ class ClientDisconnect(Exception):
 
 
 class State:
-    pass
+    def __init__(self, state_dict={}):
+        self.__dict__.update(state_dict)
```
Comment (lovelydinosaur) [inline starlette/requests.py]: This isn't right - you'll instead want to save a private `_state_dict` attribute on the instance, and read/modify it in `__getattr__`/`__setattr__`/`__delattr__`.

Comment (didip) [reply starlette/requests.py]: When I did that previously, I got infinite recursive calls. I will debug again later to figure out why.

That said, a little off-tangent here...

I found out that I can pass around singleton objects by just assigning them to app object. Example:

app.settings[‘conf’] = conf

Then I can call it with:

request.app.settings[‘conf’].

This is similar to how Tornado does it.

Comment (didip) [reply starlette/requests.py]:
```
class State(object):
    def __init__(self, state={}):
        self._state = state

    def __setattr__(self, key, item: typing.Any) -> None:
        self._state[key] = item

    def __getattr__(self, key) -> typing.Any:
        return self._state[key]

    def __delattr__(self, key) -> None:
        del self._state[key]
```

## Error
```
starlette/requests.py:23: in __init__
    self._state = state
starlette/requests.py:26: in __setattr__
    self._state[key] = item
starlette/requests.py:29: in __getattr__
    return self._state[key]
starlette/requests.py:29: in __getattr__
    return self._state[key]
E   RecursionError: maximum recursion depth exceeded while calling a Python object
```

Comment (didip) [reply starlette/requests.py]: I got it working using private `dict`:

```
class State(object):
    def __init__(self, state={}):
        self._state = state

    def __setattr__(self, key, value: typing.Any) -> None:
        if key == '_state':
            super(State, self).__setattr__(key, value)
        else:
            self._state[key] = value

    def __getattr__(self, key) -> typing.Any:
        if key == '_state':
            super(State, self).__getattr__(key)
        else:
            if key in self._state:
                return self._state[key]
            else:
                raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, key))

    def __delattr__(self, key) -> None:
        if key == '_state':
            super(State, self).__delattr__(key)
        else:
            del self._state[key]
```

I see now why you want to do this. Every state object has access to the same dict.

All tests passed, including the new ones I wrote. Not sure how the test coverage missed that 1 line...

Comment (didip) [reply starlette/requests.py]: Code coverage still stubbornly skips this line: `super(State, self).__getattr__(key)`

Even though I wrote several asserts to test the existence of keys:

```
def test_request_state_object():
    scope = {'state': {'old': 'foo'}}

    s = State(scope['state'])
    assert s._state == scope['state']
    assert getattr(s, '_state') == scope['state']

    s.new = 'value'                   # test __setattr__
    assert s.new == 'value'           # test __getattr__
    assert s._state['new'] == 'value' # test if inner _state dict is updated.

    del s.new # test __delattr__

    try:
        assert s.new == 'value' # will bombed with AttributeError
    except AttributeError as e:
        assert str(e) == "'State' object has no attribute 'new'"

    del s._state # Deleting _state dict should not bombed.
```

File: starlette/requests.py [hunk 3]
```diff
@@ -18,8 +18,30 @@ class ClientDisconnect(Exception):
...
+
+    def __setattr__(self, key, value: typing.Any) -> None:
+        if key == '_state':
+            super(State, self).__setattr__(key, value)
+        else:
+            self._state[key] = value
+
+    def __getattr__(self, key) -> typing.Any:
+        if key == '_state':
+            super(State, self).__getattr__(key)
```
Comment (lovelydinosaur) [inline starlette/requests.py]: This is the single line missing coverage ATM

Comment (didip) [reply starlette/requests.py]: Done. 100% coverage! 💯💥

**Comment by didip** [comment]
Strange, why would it fail? It completed 100%.

**Comment by lovelydinosaur** [comment]
Needs linting: https://travis-ci.org/encode/starlette/jobs/546425652#L401

Run `scripts/lint` to apply `black` and force a consistent code style everywhere.

File: starlette/requests.py [hunk 4]
```diff
@@ -18,8 +18,24 @@ class ClientDisconnect(Exception):
...
+
+    def __getattr__(self, key: typing.Any) -> typing.Any:
+        if key in self._state:
+            return self._state[key]
+        else:
+            raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, key))
+
+    def __delattr__(self, key: typing.Any) -> None:
+        if key == '_state':
+            super(State, self).__delattr__(key)
```
Comment (lovelydinosaur) [inline starlette/requests.py]: Actually we don't need to care about this case.

File: tests/middleware/test_base.py [hunk 5]
```diff
@@ -6,47 +6,46 @@
...
-        response = await call_next(request)
-        response.headers["Custom-Header"] = "Example"
-        return response
-
+def test_custom_middleware():
+    class CustomMiddleware(BaseHTTPMiddleware):
+        async def dispatch(self, request, call_next):
+            response = await call_next(request)
+            response.headers["Custom-Header"] = "Example"
+            return response
```
Comment (lovelydinosaur) [inline tests/middleware/test_base.py]: It'd be better to leave all these bits as they were - it's super important that we always try to keep the footprint of a pull request absoultely minimal, and keep it scoped to one thing only.

File: starlette/requests.py [hunk 6]
```diff
@@ -108,9 +127,10 @@ def user(self) -> typing.Any:
 
     @property
     def state(self) -> State:
-        if "state" not in self._scope:
-            self._scope["state"] = State()
-        return self._scope["state"]
+        if not hasattr(self, '_state'):
+            # Create a state instance with a reference to the dict in which it should store info
```
Comment (lovelydinosaur) [inline starlette/requests.py:130]: Let's move the `self._scope.setdefault('state', {})` to here.

File: tests/test_requests.py [hunk 7]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
...
+    assert s._state['new'] == 'value' # test if inner _state dict is updated.
+
+    del s.new # test __delattr__
+
+    try:
+        assert s.new == 'value' # will bombed with AttributeError
+    except AttributeError as e:
+        assert str(e) == "'State' object has no attribute 'new'"
+
+    del s._state # Deleting _state dict should not bombed.
```
Comment (lovelydinosaur) [inline tests/test_requests.py]: We won't need this line, once we've dropped the `__delattr__` case for `_state`.

File: tests/test_requests.py [hunk 8]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
...
+    assert getattr(s, '_state') == scope['state']
+
+    s.new = 'value'                   # test __setattr__
+    assert s.new == 'value'           # test __getattr__
+    assert s._state['new'] == 'value' # test if inner _state dict is updated.
+
+    del s.new # test __delattr__
+
+    try:
+        assert s.new == 'value' # will bombed with AttributeError
```
Comment (lovelydinosaur) [inline tests/test_requests.py]: Just something like:

```python
with pytest.assertRaises(AttributeError):
    request.state.new
```

File: tests/test_requests.py [hunk 9]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
     assert disconnected_after_response
 
 
+def test_request_state_object():
```
Comment (lovelydinosaur) [inline tests/test_requests.py:248]: I think it'd probably be marginally nicer if we tested this by instantiating a request object, and accessing `request.state` on it.

File: tests/test_requests.py [hunk 10]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
...
+    scope = {'state': {'old': 'foo'}}
+
+    s = State(scope['state'])
+    assert s._state == scope['state']
+    assert getattr(s, '_state') == scope['state']
+
+    s.new = 'value'                   # test __setattr__
+    assert s.new == 'value'           # test __getattr__
+    assert s._state['new'] == 'value' # test if inner _state dict is updated.
```
Comment (lovelydinosaur) [inline tests/test_requests.py]: Let's not test against private API.

File: tests/test_requests.py [hunk 11]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
...
 
 
+def test_request_state_object():
+    scope = {'state': {'old': 'foo'}}
+
+    s = State(scope['state'])
+    assert s._state == scope['state']
+    assert getattr(s, '_state') == scope['state']
+
+    s.new = 'value'                   # test __setattr__
```
Comment (lovelydinosaur) [inline tests/test_requests.py]: We can probably drop the "test __setattr__" style comments.

File: tests/test_requests.py [hunk 12]
```diff
@@ -240,11 +240,32 @@ def test_request_is_disconnected():
...
+
+    del s._state # Deleting _state dict should not bombed.
+
+
 def test_request_state():
     async def app(scope, receive, send):
         request = Request(scope, receive)
         request.state.example = 123
-        response = JSONResponse({"state.example": request["state"].example})
+        response = JSONResponse({"state.example": request.state.example})
```
Comment (lovelydinosaur) [inline tests/test_requests.py:271]: Yes, much nicer!

File: tests/test_websockets.py [hunk 13]
```diff
@@ -281,5 +281,5 @@ def test_websocket_scope_interface():
         send=mock_send,
     )
     assert websocket["type"] == "websocket"
-    assert dict(websocket) == {"type": "websocket", "path": "/abc/", "headers": []}
-    assert len(websocket) == 3
+    assert dict(websocket) == {"type": "websocket", "path": "/abc/", "headers": [], 'state': {}}
```
Comment (lovelydinosaur) [inline tests/test_websockets.py]: This won't have changed once we *only* set `state` to `{}` if `.state` is accessed.

File: tests/test_requests.py [hunk 14]
```diff
@@ -174,8 +174,8 @@ def test_request_scope_interface():
     """
     request = Request({"type": "http", "method": "GET", "path": "/abc/"})
     assert request["method"] == "GET"
-    assert dict(request) == {"type": "http", "method": "GET", "path": "/abc/"}
-    assert len(request) == 3
+    assert dict(request) == {"type": "http", "method": "GET", "path": "/abc/", "state": {}}
+    assert len(request) == 4
```
Comment (lovelydinosaur) [inline tests/test_requests.py:183]: This won't have changed once we only set state to {} if .state is accessed.

File: tests/middleware/test_base.py [hunk 15]
```diff
@@ -84,3 +83,41 @@ def homepage(request):
     response = client.get("/homepage")
     assert response.text == "Homepage"
     assert response.headers["Custom"] == "Example"
+
+def test_state_data_across_multiple_middlewares():
```
Comment (lovelydinosaur) [inline tests/middleware/test_base.py:89]: I'd suggest that we could simplify this down to a simpler style of test that:

* Has a single middleware, which sets some state on the request.
* Has a view, which returns `JSONResponse(dict(request.state))` 

That'd be sufficient for ensuring that `.state` is persisted across two different request instances.

**Review by lovelydinosaur — CHANGES_REQUESTED** [review]
Okay, this is *great* stuff!
I've added several review comments, for bits of cleanup. Let me knwo if anything's not clear there.
(Equally, I'm happy to pick this up and finish off the last bits if needed)

File: starlette/requests.py (hunk 6 above)
Comment (didip) [reply starlette/requests.py:130]: If I move that here, my multi middleware test failed.

File: tests/test_requests.py (hunk 8 above)
Comment (didip) [reply tests/test_requests.py]: The pytest that is part of this project doesn't seem to have `assertRaises`.

File: tests/middleware/test_base.py (hunk 15 above)
Comment (didip) [reply tests/middleware/test_base.py:89]: When I moved `"state"` init inside `@property def state():`, my test bombed in a meaningful way:

The middle state var called `bar` was not persisted.

Stacktrace:

```
_____________________________________________________________________________ test_state_data_across_multiple_middlewares ______________________________________________________________________________

    def test_state_data_across_multiple_middlewares():
        expected_value1 = "foo"
        expected_value2 = "bar"
    
        class aMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                request.state.foo = expected_value1
                response = await call_next(request)
                return response
    
        class bMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                request.state.bar = expected_value2
                response = await call_next(request)
                response.headers["X-State-Foo"] = request.state.foo
                return response
    
        class cMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                response = await call_next(request)
                response.headers["X-State-Bar"] = request.state.bar
                return response
    
        app = Starlette()
        app.add_middleware(aMiddleware)
        app.add_middleware(bMiddleware)
        app.add_middleware(cMiddleware)
    
        @app.route("/")
        def homepage(request):
            return PlainTextResponse("OK")
    
        client = TestClient(app)
>       response = client.get("/")

tests/middleware/test_base.py:122: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv/lib/python3.7/site-packages/requests/sessions.py:546: in get
    return self.request('GET', url, **kwargs)
starlette/testclient.py:421: in request
    json=json,
venv/lib/python3.7/site-packages/requests/sessions.py:533: in request
    resp = self.send(prep, **send_kwargs)
venv/lib/python3.7/site-packages/requests/sessions.py:646: in send
    r = adapter.send(request, **kwargs)
starlette/testclient.py:238: in send
    raise exc from None
starlette/testclient.py:235: in send
    loop.run_until_complete(self.app(scope, receive, send))
/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py:584: in run_until_complete
    return future.result()
starlette/applications.py:133: in __call__
    await self.error_middleware(scope, receive, send)
starlette/middleware/errors.py:122: in __call__
    raise exc from None
starlette/middleware/errors.py:100: in __call__
    await self.app(scope, receive, _send)
starlette/middleware/base.py:25: in __call__
    response = await self.dispatch_func(request, self.call_next)
tests/middleware/test_base.py:109: in dispatch
    response.headers["X-State-Bar"] = request.state.bar
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <starlette.requests.State object at 0x10ab52e48>, key = 'bar'

    def __getattr__(self, key: typing.Any) -> typing.Any:
        if key in self._state:
            return self._state[key]
        else:
            raise AttributeError(
>               "'{}' object has no attribute '{}'".format(self.__class__.__name__, key)
            )
E           AttributeError: 'State' object has no attribute 'bar'

starlette/requests.py:33: AttributeError
```

File: starlette/requests.py [hunk 16]
```diff
@@ -18,8 +18,24 @@ class ClientDisconnect(Exception):
...
+    def __getattr__(self, key: typing.Any) -> typing.Any:
+        if key in self._state:
+            return self._state[key]
+        else:
+            raise AttributeError(
+                "'{}' object has no attribute '{}'".format(self.__class__.__name__, key)
+            )
+
+    def __delattr__(self, key: typing.Any) -> None:
+        if hasattr(self, "_state"):
```
Comment (lovelydinosaur) [inline starlette/requests.py]: I don't think we need the `if` condition here.

**Comment by lovelydinosaur** [comment]
Great stuff! Followed up and completed this work in #557

**Comment by didip** [comment]
Awesome! Thank you for helping to get it merged.

**Pull request closed without merging**